cmds/exp/disk_unlock/disk_unlock.go

```diff
@@ -166,18 +59,6 @@ func writeFile(filename string, contents string) error {
 func main() {
```

Comment:
To make the code more testable in general we've been renaming `main` functions to `run` and passing in explicit dependencies (example: https://github.com/u-root/u-root/pull/2526/files).  Can you please consider restructuring the code and adding tests? Thanks!

Reply:
I made this change for `nvme_unlock`, but for `disk_unlock` I would prefer to make as few changes as possible, since I don't have a good way of testing it for regressions at the moment.

For testing, I'm not sure if there's a good way to unit test much of the functionality here.  I could add a few tests for `GenPassword`, but aside from that, nearly all the code is dealing with interacting with the hardware on the system (IPMI, Opal driver, etc).  I assume an integration test would be more appropriate, but I don't know if there's a reasonable way to fake those external components in qemu.

Reply:
> I would prefer to make as few changes as possible, since I don't have a good way of testing it for regressions at the moment.

This screams "we need test coverage for our code, otherwise people in the future may break this and we won't know" :(

I wonder if this can be made more testable by re-arranging the binaries, packages and abstractions. Can you think of a way to structure the code such that the system level stuff is abstracted and can be mocked out with interfaces or dependency injection?

Reply:
I agree, but refactoring this code for unit tests would take some pretty invasive changes, since you'd need to mock out the filesystem, the key generation function, the SCSI driver interface, and the block device driver interface.  It's not impossible but the cost/benefit is questionable.

Either way, it feels out of scope for this PR, since this file is only touched for the sake of moving some of its functions out into a shared package.

Reply:
We have the ability to run Go unit tests in QEMU VMs as root, so we don't need to mock all that out. But I do agree it's out of scope for this PR.

Reply:
Agreed that it's not realistic to ask to mock out the SCSI interface, block driver interface, filesystem calls etc.   However, there's a lot of code, for example in disk_unlock.go that can be tested that doesn't have to deal with the hardware - code which can be refactored to be made testable with dependency injection, interfaces etc.

For example, all the code handling the security statuses (https://github.com/u-root/u-root/blob/b250f8cc60d0605c79265383697e2af53a89e61f/cmds/exp/disk_unlock/disk_unlock.go#L87-L110), all of the code handling retrying multiple keys (https://github.com/u-root/u-root/blob/b250f8cc60d0605c79265383697e2af53a89e61f/cmds/exp/disk_unlock/disk_unlock.go#L114-L129), etc.

Is it reasonable to ask that the code in the file be re-organized such that it is more testable without having to mock out system level stuff?